design-proposals/template-generation.md

```diff
@@ -0,0 +1,131 @@
+
+The controller watching `VirtualMachineTemplateRequest` objects will see the request and post a template of the request VirtualMachine in the specified format.
+    
+## VirtualMachineTemplateRequest API
+
+The VirtualMachineTemplateRequest specification fields:
+
+* **VMName**: the name of a VirtualMachine living in the cluster which will be the reference from which the template will be created (required field).
+* **VMNamespace**: the namespace of the reference VM. If not specified the VM will be looked up in the namepace in which VirtualMachineTemplateRequest was created (optional field).
+* **TemplateName**: the name of the template object that will be generated, if omitted it will be auto-generated (otional field).
+* **Format**: the format of the output template (required field).
+
+The first implemented format will be the [OpenShift Template(https://docs.openshift.com/container-platform/4.9/openshift_images/using-templates.html#templates-writing_using-templates)] format.
+
+In order to support multiple formats, the controller will have to be extended to generate the correct output for those formats.
+
+## API Examples
+
+Given the following VM
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-ephemeral
+spec:
+  template:
+    domain:
+      devices:
+        disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      resources:
+        requests:
+          memory: 128Mi
+    volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/cirros-container-disk-demo:devel
+      name: containerdisk
+```
+
+After posting this VirtualMachineTemplateRequest object
+
+```yaml
+apiVersion: ssp.kubevirt.io/v1alpha1
+kind: VirtualMachineTemplateRequest
+metadata:
+  name: virtualmachinetemplaterequest-sample
+spec:
+  vmName: vmi-ephemeral
+  templateName: vmi-ephemeral-template
+  format: template.openshift.io/v1
+```
+
+The following template will be created
+
+```yaml
+apiVersion: v1
+kind: Template
+metadata:
+  name: vmi-ephemeral-template
+  labels: ${LABELS}
+objects:
+- apiVersion: kubevirt.io/v1
+  kind: VirtualMachine
+  metadata:
+    generateName: vmi-ephemeral
+  template:
+    spec:
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: ${DISK0_BUS}
+            name: containerdisk
+        resources:
+          requests:
+            memory: ${REQUESTS_MEMORY}
+      volumes:
+      - containerDisk:
+          image: ${VOLUME0_IMAGE}
+        name: containerdisk
+parameters:
+- description: disk0-bus
+  name: DISK0_BUS
+  value: virtio
+- description: requests-memory
+  name: REQUESTS_MEMORY
+  value: 128Mi
+- description: volume0-image
+  name: VOLUME0_IMAGE
+  value: registry:5000/kubevirt/cirros-container-disk-demo:devel
+```
+
+## Template generation
+
+The template generation mechanism follows some basic rules:
```

Comment:
We recently added flavors to core kubevirt API. What is your vision of the flavor in your design? (Can we use them for possible suggested parameters?)